### PR TITLE
Comment (and cleanup) group by tests

### PR DESCRIPTION
**Pull request by dt**
Adding comments to explain purpose of some tests
Re-ordered some tests to try to improve the narrative flow from test to test.
Removed some duplicative tests around GROUP BY.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3656)

### CODE REVIEW

File: sql/testdata/aggregate [hunk 1]
```diff
@@ -165,33 +211,16 @@ SELECT MAX(k), MIN(v) FROM kv HAVING MAX(v) > 2;
...
+query error column "k" must appear in the GROUP BY clause or be used in an aggregate function
+SELECT 3 FROM kv GROUP BY v HAVING k > 5;
 
+# pg has a special case for grouping on primary key, which would allow this, but we do not.
+# See http://www.postgresql.org/docs/current/static/sql-select.html#SQL-GROUPBY
 query error column "v" must appear in the GROUP BY clause or be used in an aggregate function
 SELECT 3 FROM kv GROUP BY k HAVING v > 2
 
 query error column "k" must appear in the GROUP BY clause or be used in an aggregate function
-SELECT 3 FROM kv GROUP BY v HAVING k > 5;
```
Comment (tamird) [inline sql/testdata/aggregate]: keep this one. if someone decides to implement the special case above, they're gonna need this test.

File: sql/testdata/aggregate [hunk 2]
```diff
@@ -165,33 +211,16 @@ SELECT MAX(k), MIN(v) FROM kv HAVING MAX(v) > 2;
...
-----
-3
-
-query I rowsort
-SELECT 3 FROM kv WHERE k > 5;
-----
-3
-3
-3
+# Columns listed in the HAVING must conform to same validation as SELECT (grouped or aggregated)
```
Comment (tamird) [inline sql/testdata/aggregate]: s/Columns/Expressions/ s/HAVING/HAVING clause/ s/SELECT/SELECT clause/

File: sql/testdata/aggregate [hunk 3]
```diff
@@ -50,89 +72,113 @@ SELECT COUNT(*), k FROM kv GROUP BY -4
...
+----
+1 12
+1 3
+1 7
+1 8
+1 9
+1 NULL
+
+
+# Selecting a more complex expression, made of of things which are each grouped, works.
```
Comment (tamird) [inline sql/testdata/aggregate]: s/of of/up of/

File: sql/testdata/aggregate [hunk 4]
```diff
@@ -50,89 +72,113 @@ SELECT COUNT(*), k FROM kv GROUP BY -4
...
 1 NULL
 2 A
 2 B
 
-
+# Selecting a value that is not grouped, even if a function of it it, does not work.
 query error column "s" must appear in the GROUP BY clause or be used in an aggregate function
 SELECT COUNT(*), s FROM kv GROUP BY UPPER(s)
 
+# Selecting and grouping on a more complex expression works as well.
```
Comment (tamird) [inline sql/testdata/aggregate]: s/as well//

File: sql/testdata/aggregate [hunk 5]
```diff
@@ -50,89 +72,113 @@ SELECT COUNT(*), k FROM kv GROUP BY -4
...
-SELECT COUNT(*), UPPER(kv.s) FROM kv GROUP BY s
+SELECT COUNT(*), UPPER(s) FROM kv GROUP BY UPPER(s)
 ----
-1 A
 1 NULL
-2 A
 2 B
+3 A
 
+# Selecting a function of something which is grouped also works.
```
Comment (tamird) [inline sql/testdata/aggregate]: s/also //

File: sql/testdata/aggregate [hunk 6]
```diff
@@ -28,6 +49,7 @@ SELECT COUNT(*), k FROM kv GROUP BY k
 1 7
 1 8
 
+# Exactly the same as previous except GROUP BY specified using column number.
```
Comment (tamird) [inline sql/testdata/aggregate]: i would avoid referring to previous test cases like this. restate the context.

File: sql/testdata/aggregate [hunk 7]
```diff
@@ -10,11 +10,32 @@ statement OK
...
+
+# Even with no aggregate functions, grouping occurs.
+query I rowsort
+SELECT 1 FROM kv GROUP BY v
+----
+1
+1
+1
+
+# Presence of HAVING triggers aggregation, reducing one row (even without GROUP BY).
```
Comment (tamird) [inline sql/testdata/aggregate]: s/reducing/producing/? returning?

File: sql/testdata/aggregate [hunk 8]
```diff
@@ -10,11 +10,32 @@ statement OK
...
 (8, 4, 2, 'A')
 
+
+# Aggregate expression triggers aggregation (thus 1 row) even if it simplifies to a constant.
+query I rowsort
+SELECT MIN(1) FROM kv
+----
+1
+
+# Even with no aggregate functions, grouping occurs.
```
Comment (tamird) [inline sql/testdata/aggregate]: s/occurs/occurs in the presence of GROUP BY/

**Comment by tamird** [comment]
LGTM. s/group by/GROUP BY/ in commit message, though.

**Comment by dt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/testdata/aggregate, line 25 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-25:-K7NxXkmJlaG42zxr_qq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L25)):*
Done.

---

*[sql/testdata/aggregate, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-33:-K7NxXIbiB3el94wwyF8:173817976) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L33)):*
`reducing results to` ?

---

*[sql/testdata/aggregate, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-52:-K7NxlfTzA5rnX66qNrf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L52)):*
Done.

---

*[sql/testdata/aggregate, line 136 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-136:-K7Nxo1Gp573TXiQwW9a:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L136)):*
Done.

---

*[sql/testdata/aggregate, line 149 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-149:-K7NxrWA9k0Jp8S2LUEq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L149)):*
Done.

---

*[sql/testdata/aggregate, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-161:-K7Nxu85IutCczW3NECP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L161)):*
Done.

---

*[sql/testdata/aggregate, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-214:-K7Ny2W154QSTj3YzsY8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L214)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/testdata/aggregate, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-52:-K7NzbWLFWCaskSp_pZE:228445400) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L52)):*
Hm. How about "GROUP BY specified using column index works"?

---

*[sql/testdata/aggregate, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-214:-K7NzlgmmYNoPEaM2FCI:-1513982435) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L214)):*
um, this got mangled and is missing articles

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656)*

**Comment by dt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/testdata/aggregate, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-52:-K7OCN7txILB6TVYsJzn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L52)):*
Done.

---

*[sql/testdata/aggregate, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656#-K7NvrR8vThMDIZtn5X_-r1-214:-K7OCKlMZhegyA1N03V8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/31bc3797b3abb7f6feb7c808e0cc52995f367c87/sql/testdata/aggregate#L214)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3656)*

**Pull request merged**